CONTRIBUTING.md

```diff
@@ -0,0 +1,63 @@
+
+For useful resources, please see:
+
+- [Git basics](https://git-scm.com/book/en/v1/Getting-Started-Git-Basics) for Git beginners
+- [Git best practices](https://sethrobertson.github.io/GitBestPractices/)
+
+## Submitting changes (Pull Requests)
+
+If you are a complete newbie, click [here](https://github.com/firstcontributions/first-contributions) for an easy-to-follow guide (with pictures!)
+
+We follow the standard procedure for submitting Pull Requests. Please refer to the [official GitHub documentation](https://help.github.com/articles/creating-a-pull-request/) if you are unfamiliar with the procedure. If you still need help, we are more than happy to guide you along!
+
+**Note:** Before submitting code, please either choose an existing Issue, or write your own, describing the problem you are solving. We would like for every PR to have an accompanying Issue, so that we know concretely what the problem is and can track its resolution.
+
+### Code Style
+
+We try to follow standard best practices for every language and framework we use. These best practices are enforced by CI (see below) and code review (also see below).
+
+### Format
+
+We don't have any hard-and-fast rules for styling git commits or branch names, but try to look at existing PRs first to get a sense of how they are structured.
+
+### Running tests (CI script)
+
+Submitted PRs are expected to pass continuous integration (CI), which, among other things, runs a test suite on your PR to make sure that your code doesn't have any bugs.
+
+Please refer to the README of this repo for instructions on running the CI and test suites locally.
+
+We require that every piece of code is _covered_, meaning tested. If you aren't familiar with writing tests for the language, make your PR anyway and we'll help you from there.
+
+### Code review
+
+Your PR will be assigned to a team member and reviewed promptly. More often than not, a code submission will be met with review comments and changes requested. Keep in mind that it's nothing personal -- we leave each other review comments all the time.
+
+After addressing review comments, it is up to your discretion whether you make a new commit or amend the previous one and do a [force-push](https://estl.tech/a-gentler-force-push-on-git-force-with-lease-fb15701218df) to the PR branch. We encourage amending the last commit if there are only minor changes to be made. GitHub now shows diffs for force-pushed commits, making them easy to review, and it keeps the final commit history clean.
```

Comment:
IMO, force-push should be discouraged as much as possible. If we want a clean git history we can squash PRs on merge. If someone force-pushes 5 minor lines in a 500 line commit we don't have a good way of seeing just those 5 lines and we need to go over the entire commit again. I'm not convinced that's the best approach.

Reply:
> If someone force-pushes 5 minor lines in a 500 line commit we don't have a good way of seeing just those 5 lines and we need to go over the entire commit again.

GitHub shows diffs for force-pushes since a bit over a year ago so you would only see those 5 lines.

We don't actually have an official policy on force-push and so far people have just been doing what they like, and it's been working. I'll update this section to say that it's personal preference.